File: 0057-insert-interval/0057-insert-interval.py
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        
        output = []
        i = 0
        
        while i < len(intervals) and intervals[i][1] < newInterval[0]:
            output.append(intervals[i])
            i += 1
        if output and output[-1][-1] >= newInterval[0]:
            output[-1] = [min(output[-1][0], newInterval[0]), max(output[-1][-1], newInterval[1])]
        else:
            output.append(newInterval)

        while i < len(intervals):
            start, end = intervals[i]
            if start <= output[-1][-1]:
                output[-1] = [min(output[-1][0], start), max(output[-1][-1], end)]
            else:
                output.append(intervals[i])
            i += 1
                
        return output
                          

File: 0057-insert-interval/0057-insert-interval.cpp
class Solution {
public:
    vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {

        vector<vector<int>> output;
        int i = 0;

        while (i < intervals.size() and intervals[i][0] < newInterval[0]) {
            output.emplace_back(intervals[i]);
            i++;
        }

        // Insert newInternval and check if merge
        if (!output.empty() and output.back()[1] >= newInterval[0]) {
            output.back()[1] = max(output.back()[1], newInterval[1]);
        } else {
            output.emplace_back(newInterval);
        }

        // Add remaining intervals
        while (i < intervals.size()) {
            if (output.back()[1] >= intervals[i][0]) {
                output.back()[1] = max(output.back()[1], intervals[i][1]);
            } else {
                output.emplace_back(intervals[i]);
            }
            i++;
        }
        
        return output;
    }
};